### PR TITLE
Unified error handling in truss server and adding baseten error headers to response

### PR DESCRIPTION
**Pull request by marius-baseten**
## :rocket: What
* Unify error handling and add baseten custom error headers to responses [spec](https://www.notion.so/ml-infra/End-to-end-Error-code-Propagation-698daa6d2c3a4d10b55056b1a892cdfb).
* Remove dead code (unused exceptions and error handlers).
* Fuse optional parsing of pydantic models ("truss schema") with top-level input deserialization. This allows to parse json with pydantic directly (`Model(**json.loads(x))` => `Model.parse_raw(x)`), which is more performant than using `json.loads` first.

See also: https://github.com/basetenlabs/baseten/pull/9258

### CODE REVIEW

File: truss/templates/server/common/truss_server.py [hunk 1]
```diff
@@ -145,20 +141,39 @@ async def predict(
             if self.is_binary(request):
                 with tracing.section_as_event(span, "binary-deserialize"):
                     body = truss_msgpack_deserialize(body_raw)
+                if model.truss_schema:
+                    try:
+                        with tracing.section_as_event(span, "parse-pydantic"):
+                            body = model.truss_schema.input_type.parse(**body)
```
Comment (squidarth) [inline truss/templates/server/common/truss_server.py:147]: non-blocking, and I also don't think we have very many examples of this anyway, but would be good to have a test in test_model_inference for this case (binary w pydantic typing)

File: truss/templates/server/common/errors.py [hunk 2]
```diff
@@ -57,44 +37,121 @@ def __str__(self):
...
+
+@overload
+def intercept_exceptions(
+    func: Callable[_P, _R_async], logger: logging.Logger
+) -> Callable[_P, _R_async]: ...
+
+
+def intercept_exceptions(
+    func: Callable[_P, _R], logger: logging.Logger
+) -> Callable[_P, _R]:
```
Comment (squidarth) [inline truss/templates/server/common/errors.py:135]: nice typing

File: truss/templates/server/common/errors.py [hunk 3]
```diff
@@ -57,44 +37,120 @@ def __str__(self):
...
-        content={"error": f"{type(exc).__name__} : {str(exc)}"},
-    )
-
-
-async def model_not_found_handler(_, exc):
-    return JSONResponse(status_code=HTTPStatus.NOT_FOUND, content={"error": str(exc)})
+def _make_baseten_response(
+    http_status: int,
+    info: Union[str, Exception],
+    baseten_error_code: Optional[int] = None,
```
Comment (squidarth) [inline truss/templates/server/common/errors.py:61]: is the plan to start passing this argument later?

**Review by squidarth — APPROVED** [review]

File: truss/templates/server/common/errors.py (hunk 3 above)
Comment (marius-baseten) [reply truss/templates/server/common/errors.py:61]: Yes, the concrete codes are still up for review with @nicolasgere .

File: truss/templates/server/common/truss_server.py (hunk 1 above)
Comment (marius-baseten) [reply truss/templates/server/common/truss_server.py]: Good point, that actually helped find a bug!

**Review by nicolasgere — APPROVED** [review]

**Pull request merged**